src/java.base/share/classes/java/util/ImmutableCollections.java

```diff
@@ -1357,6 +1387,52 @@ private Object writeReplace() {
             }
             return new CollSer(CollSer.IMM_MAP, array);
         }
+
+        @Override
+        public Set<K> keySet() {
+            return new AbstractSet<>() {
+                public Iterator<K> iterator() {
+                    return new Iterator<>() {
+                        private final Iterator<Entry<K, V>> i = new MapNIterator();
+                        public boolean hasNext() { return i.hasNext(); }
+                        public K next() { return i.next().getKey(); }
+                        public void remove() { throw uoe(); }
+                    };
+                }
+                public int size() { return MapN.this.size; }
+                public boolean isEmpty() { return size() == 0; }
+                public void clear() { throw uoe(); }
+                public boolean contains(Object k) { return MapN.this.containsKey(k); }
+            };
+        }
+
+        @Override
+        public Collection<V> values() {
+            return new AbstractCollection<>() {
+                public Iterator<V> iterator() {
+                    return new Iterator<>() {
+                        private final Iterator<Entry<K,V>> i = new MapNIterator();
+                        public boolean hasNext() { return i.hasNext(); }
+                        public V next() { return i.next().getValue(); }
+                        public void remove() { throw uoe(); }
+                    };
+                }
+                public int size() { return MapN.this.size;}
+            };
+        }
+
+        @Override
+        public void forEach(BiConsumer<? super K, ? super V> action) {
+            for (int i = 0; i < table.length; i += 2) {
```

Comment:
This does not iterate in the same order as the iterator. See #15834 for a correct one (which is somehow 4% slower than default impl in JMH after inlining)

Also missing null check if this map is empty.